### ISSUE TITLE
constraint solver

### ISSUE THREAD

**Issue by ccorcos** [facebook/css-layout#56]
Have you guys though of building in a constraint solver? I believe this is how Apple does layout. This way you can do things like set two elements equidistant from the center, or set the distance between to items. I think this would really take the web to the next level.

https://www.npmjs.com/package/cassowary

**Comment by sidorares** [comment]
https://twitter.com/Vjeux/status/561990962071625728

**Comment by ccorcos** [comment]
You can't express a boolean operation with linear inequalities -- correct. But you can compute the boolean operations and then do the constraint solver on what's left...

**Comment by ccorcos** [comment]
I see what you're getting at here with text-wrapping... the problem is the order of operations. If you know the width, you can compute the height (with text wrapping). But if you don't know the width yet, then you can't compute the height yet, and everything is fucked. is that what you're talking about?

**Comment by ccorcos** [comment]
It seems like this could be possible. So long as the width isn't based on the height. You could solve everything up to the width. The calculate the height, then calculate the rest. Anything that depends on the height will need to wait to be calculated...

**Comment by vjeux** [comment]
> You could solve everything up to the width. The calculate the height, then calculate the rest.

Yeah, this is what css-layout implementation does. The issue is that it's not how the constraint solver works. You can't really inject rules dynamically like this. You need to know up-front what the constraints are.

If you don't, then you need to do multiple passes by using a fake height, but then it's slow because multiple passes, and you don't have a guarantee that it's going to converge anymore.

GSS last time I checked does only one pass and fix-up the height of text afterwards, but it's no longer guaranteed to be the optimal layout :(

**Comment by ccorcos** [comment]
If your problem is with text-height, you can set the default height to something like 1. Then, since the width CANNOT depend on the height, you can calculate the new height for all of them. One more pass with these heights, and you should have the layout, right?

**Comment by vjeux** [comment]
If you have multiline text, the height MUST depend on the width.

**Comment by ccorcos** [comment]
yes, but not the other way around. there dependency must be one-way. height depends on width. width doesnt depend on height. so set the height to some constant. Calculate everything else. Then calculate the heights based on the computed widths, etc.

**Comment by vjeux** [comment]
the problem is that many other things depend on the height of the text. The height of the container, the top of the next element...

**Comment by RangerMauve** [comment]
Is two passes really that bad a solution?

**Comment by vjeux** [comment]
You may have other constraints that are invalidated by this second pass. So you need a third ... and maybe a fourth and so on. And, because you are using heuristics, you lose the convergence property of the constraint solver. So you may never find a proper solution

**Comment by ccorcos** [comment]
container height is a perfect counterexample. let me think on this a little more...

**Comment by ccorcos** [comment]
So image this scenario where we have two text elements on top of each other. We set their heights to a constant. Run the solver to get their widths. Then we rerun the solver to get their heights. This scenario works right?

Ok. Now I'm thinking your container counterexample does disprove the point. If we set the height of the text elements to a constant, we may have all kinds of wrong values left over after we do the constraint solver (the container height, etc.). But the values that we can be sure are correct are the text element widths because the text width cannot depend on the text height. Thus, if we run the constraint solver one more time with the proper text element heights, then the container height will be adjusted and everything else should work out just fine. 

So yeah, two passes should work. right?
